### ISSUE TITLE
Authy.download fails

### ISSUE THREAD

**Issue by grahampugh** [autopkg/nzmacgeek-recipes#9]
Unfortunately your `Authy.download` recipe is failing due to a `curl` error:

```
% autopkg run -v Authy.download       
Processing Authy.download...
URLTextSearcher
URLTextSearcher: Found matching text (url_to_download): https://electron.authy.com/download?channel=stable&arch=x64&platform=darwin&version=latest&product=authy
URLDownloader
Command '['/usr/bin/curl', '--silent', '--show-error', '--no-buffer', '--dump-header', '-', '--speed-time', '30', '--location', '--url', 'https://electron.authy.com/download?channel=stable&arch=x64&platform=darwin&version=latest&product=authy', '--fail', '--output', '/Users/gpugh/Library/AutoPkg/Cache/local.download.Authy/downloads/tmp6mrgu6tw']' returned non-zero exit status 22.
Failed.
```

If the `--location` flag is removed, the curl command will succeed, but in discussion with @nmcspadden this is not possible without building out your own processor. He says you can supply a raw curl command which can be read by the `URLGetter` from within your own processor, but I haven't had time to figure out how to do this - it doesn't seem to be documented.

**Comment by nmcspadden** [comment]
It is indeed documented: https://github.com/autopkg/autopkg/wiki/Downloading-from-the-Internet-in-Custom-Processors#bring-your-own-building-blocks-with-selfdownload_with_curlcurl_cmd

**Comment by grahampugh** [comment]
I decided to take a stab at this, and I was barking slightly up the wrong tree. The problem is that the URL that is grabbed during the redirect contains a space:

```
https://s3.amazonaws.com/authy-electron-repository-production/authy/stable/1.8.3/darwin/x64/Authy Desktop-1.8.3.dmg
```

`curl` seemingly cannot handle this, although Safari can. 

So, I created a new processor `URLDownloaderForAuthy`, which is a copy-paste of the standard `URLDownloader` processor, and functions the same by default. But, I added an additional key `prefetch_and_encode_redirect_url`, which, when True, pre-fetches the redirect URL and encodes it, before continuing on to get the content of the redirected URL. This is achieved largely through a new function, also with the name `prefetch_and_encode_redirect_url`.

I could attempt to PR `URLDownloader` itself to add this functionality, but I'm not sure if it would be accepted. Anyway, it solves the problem for Authy.

**Comment by nmcspadden** [comment]
That feels like the wrong solution to me. The URL should be escaped before it gets handled, which implies that either there's a bug in the URL collation happening in URLDownloader, or that something about the download recipe is passing in a value at the wrong time.

**Comment by grahampugh** [comment]
I guess it's an edge case that a URL would contain a space, more so the redirect URL, not the one supplied by the recipe. News to me that a URL could even have a space in it! Maybe it's a curl headers artefact. 

But, I don't see another way to fix the problem when it's the URL being obtained by curl's redirect_url that has the space in it. But of course, open to ideas and especially open to it being solved in URLDownloader itself.

**Comment by nmcspadden** [comment]
URLs can't actually contain spaces; the browsers automatically convert spaces to %20. Looks like Authy's redirect to Amazon assumes it will _always_ come from a browser that does that for them, because they're being lazy. 

I think it makes more sense to add the escaping of URLs a bit farther up the chain in URLDownloader, then...

**Comment by moorereason** [comment]
Upstream issue fixed in curl 7.67.0: https://github.com/curl/curl/issues/4445

**Comment by grahampugh** [comment]
Unfortunately, we have to work with what Apple give us:
```
gpugh@Grahams-M1-MacBook-Air ~ % curl --version
curl 7.64.1 (x86_64-apple-darwin20.0) libcurl/7.64.1 (SecureTransport) LibreSSL/2.8.3 zlib/1.2.11 nghttp2/1.41.0
Release-Date: 2019-03-27
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS GSS-API HTTP2 HTTPS-proxy IPv6 Kerberos Largefile libz MultiSSL NTLM NTLM_WB SPNEGO SSL UnixSockets
```

**Comment by moorereason** [comment]
> Unfortunately, we have to work with what Apple give us:

Agreed.  I just wanted to share that the root cause of the issue was in the client.  As a server guy, I couldn't let the accusation of laziness on the server side go unanswered.  :smile:

**Comment by PeetMcK** [comment]
This can be fixed by using the brew version of curl ... 

`<key>CURL_PATH</key>
				<string>/usr/local/Cellar/curl/7.75.0/bin/curl</string>`

Or scraping the url from from the appcast url in the macupdater.net [authy.rb](https://www.macupdater.net/casks_manual/authy.rb)

It would seem that the folk(s) at corecode.io have written the code necessary to clean spaces and other nasty bits from poorly constructed redirects. 

`https://macupdater.net/cgi-bin/check_urls/check_url_redirect.cgi?url=https://electron.authy.com/download%3Fchannel=stable%26arch=x64%26platform=darwin%26version=latest%26product=authy`

**Comment by PeetMcK** [comment]
> > Unfortunately, we have to work with what Apple give us:
> 
> Agreed. I just wanted to share that the root cause of the issue was in the client. As a server guy, I couldn't let the accusation of laziness on the server side go unanswered. 😄

To be clear though, this isn't technically a curl issue. It's an issue with Authy making a lazy assumption and providing an invalid url with a space in it. The curl project looks to have given up and allowed now allows malformed url's that have spaces.

**Comment by grahampugh** [comment]
I still feel that the only short-term solution is to use a custom processor such as the one I have included in my PR #10. I guess I can use my own fork in the meantime.

**Comment by PeetMcK** [comment]
I saw your [PR #10](https://github.com/autopkg/nzmacgeek-recipes/pull/10) ... and boy does it make me feel like a lazy admin.  I agree that a custom processor is the "right" way to go, but unless your changes were folded into the primary `autopkg` `URLDownloader` tying this a Authy Desktop download recipe to an up-to-date curl from brew is easier on my brain. For me it's less overhead and and a current curl is easier to maintain than yet another processor.

All that said, bravo, on the processor. Considering 10.12.1 has a curl build from March 2019, the functionality you added should probably get into URLDownloader as we're likely not going to see curl 7.67.0+ for some time.

**Comment by grahampugh** [comment]
I don't want to install brew on my AutoPkg server :) And, that would be unscalable for those people with ephemeral AutoPkg runners.

I'm not sure whether it's worth editing `URLDownloader` for apparently one broken recipe due to lazy app development, especially when there is a prospect of it being fixed via an update to `curl`. 

I guess another alternative, which would improve the predictability of AutoPkg, would be to bundle a known version of `curl` in with the AutoPkg package. @nmcspadden any comments on that?

**Comment by PeetMcK** [comment]
Or barring a the overhead of bundling a known `curl`, perhaps a `CURL_PATH` like the `GIT_PATH` in the `com.github.autopkg` preference domain?

**Comment by grahampugh** [comment]
I note that my fork is no longer successfully downloading the DMG. This recipe is therefore currently dead in all its forms, with no obvious fix. I've reached out to Twilio support to see if they have a solution such as an alternative URL.

**Comment by fuzzylogiq** [comment]
Any luck with that?
